core/src/consensus.rs

```diff
@@ -606,6 +703,203 @@ impl Tower {
         bank_weight
     }
 
+    fn voted_slots(&self) -> Vec<Slot> {
+        self.lockouts
+            .votes
+            .iter()
+            .map(|lockout| lockout.slot)
+            .collect()
+    }
+
+    pub fn is_stray_last_vote(&self) -> bool {
+        if let Some(last_voted_slot) = self.last_voted_slot() {
+            if let Some(max_stray_slot) = self.stray_restored_slots.iter().next_back() {
+                return *max_stray_slot == last_voted_slot;
+            }
+        }
+
+        false
+    }
+
+    fn stray_restored_ancestors(&self) -> HashSet<Slot> {
+        let mut ancestors: HashSet<_> = self.stray_restored_slots.clone().into_iter().collect();
+        if let Some(last_voted_slot) = self.last_voted_slot() {
+            assert!(ancestors.contains(&last_voted_slot));
+            ancestors.remove(&last_voted_slot);
+        }
```

Comment:
This subtle handling is needed to be the `ancestors` of `last_voted_slot`. Also, `self.stray_restored_slots` must contain `last_voted_slot` because of `is_stray_last_vote`. (Btw, this is the reason of not being called `stray_restored_ancestors`) Otherwise, we can't construct that information...